miniapps/shifted/extrapolator.hpp

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_EXTRAPOLATOR_HPP
+#define MFEM_EXTRAPOLATOR_HPP
+
+#include "mfem.hpp"
+
+namespace mfem
+{
+
+class DiscreteUpwindLOSolver;
+class FluxBasedFCT;
+
+class AdvectionOper : public TimeDependentOperator
+{
+private:
+   Array<bool> &active_zones;
+   ParBilinearForm &M, &K;
+   HypreParMatrix *K_mat;
+   const Vector &b;
+
+   DiscreteUpwindLOSolver *lo_solver;
+   Vector *lumpedM;
+
+   void ComputeElementsMinMax(const ParGridFunction &gf,
+                              Vector &el_min, Vector &el_max) const;
+   void ComputeBounds(const ParFiniteElementSpace &pfes,
+                      const Vector &el_min, const Vector &el_max,
+                      Vector &dof_min, Vector &dof_max) const;
+   void ZeroOutInactiveZones(Vector &dx);
+
+public:
+   // HO is standard FE advection solve; LO is upwind diffusion.
+   enum AdvectionMode {HO, LO} adv_mode = AdvectionOper::HO;
+
+   AdvectionOper(Array<bool> &zones, ParBilinearForm &Mbf,
+                 ParBilinearForm &Kbf, const Vector &rhs);
+
+   ~AdvectionOper();
+
+   virtual void Mult(const Vector &x, Vector &dx) const;
+};
+
+// Extrapolates through DG advection based on:
+// [1] Aslam, "A Partial Differential Equation Approach to Multidimensional
+// Extrapolation", JCP 193(1), 2004.
+// [2] Bochkov, Gibou, "PDE-Based Multidimensional Extrapolation of Scalar
+// Fields over Interfaces with Kinks and High Curvatures", SISC 42(4), 2020.
+class Extrapolator
+{
+public:
+   enum XtrapType {ASLAM, BOCHKOV} xtrap_type = ASLAM;
+   AdvectionOper::AdvectionMode advection_mode = AdvectionOper::HO;
+   int xtrap_order    = 1;
+   bool visualization = false;
+   int vis_steps      = 5;
+
+   Extrapolator() { }
+
+   // The known values taken from elements where level_set > 0, and extrapolated
+   // to all other elements. The known values are not changed.
+   void Extrapolate(Coefficient &level_set, const ParGridFunction &input,
```

Comment:
Can we make the time_period input as const double instead of just double?

Reply:
Done.